src/java.base/share/classes/java/lang/template/StringTemplateImplFactory.java

```diff
@@ -0,0 +1,205 @@
+        valuesMH = MethodHandles.filterArguments(valuesMH, 0, components);
+        mt = methodType(List.class, StringTemplateImpl.class);
+        valuesMH = MethodHandles.permuteArguments(valuesMH, mt, permute);
+
+        MethodHandle constructor = MethodHandles.insertArguments(CONSTRUCTOR,0,
+                elements.primitiveCount(), elements.objectCount(),
+                fragments, valuesMH, interpolateMH);
+        constructor = MethodHandles.foldArguments(elements.initializer(), 0, constructor);
+
+        mt = methodType(StringTemplate.class, ptypes);
+        constructor = constructor.asType(mt);
+
+        return constructor;
+    }
+
+    /**
+     * Generic {@link StringTemplate}.
+     *
+     * @param fragments  immutable list of string fragments from string template
+     * @param values     immutable list of expression values
+     */
+    private record SimpleStringTemplate(List<String> fragments, List<Object> values)
+            implements StringTemplate {
+        @Override
+        public String toString() {
+            return StringTemplate.toString(this);
+        }
+    }
+
+    /**
+     * Returns a new StringTemplate composed from fragments and values.
+     *
+     * @param fragments array of string fragments
+     * @param values    array of expression values
+     *
+     * @return StringTemplate composed from fragments and values
+     */
+    @Override
+    public StringTemplate newStringTemplate(String[] fragments, Object[] values) {
+        return new SimpleStringTemplate(List.of(fragments), toList(values));
+    }
+
+    /**
+     * Returns a new StringTemplate composed from fragments and values.
+     *
+     * @param fragments list of string fragments
+     * @param values    array of expression values
+     *
+     * @return StringTemplate composed from fragments and values
+     */
+    @Override
+    public StringTemplate newStringTemplate(List<String> fragments, Object[] values) {
+        return new SimpleStringTemplate(List.copyOf(fragments), toList(values));
+    }
+
+    /**
+     * Returns a new StringTemplate composed from fragments and values.
+     *
+     * @param fragments list of string fragments
+     * @param values    list of expression values
+     *
+     * @return StringTemplate composed from fragments and values
+     */
+    @Override
+    public StringTemplate newStringTemplate(List<String> fragments, List<?> values) {
+        return new SimpleStringTemplate(List.copyOf(fragments), toList(values.stream().toArray()));
+    }
+
+    /**
+     * Collect nullable elements from an array into a unmodifiable list.
+     *
+     * @param elements  elements to place in list
+     *
+     * @return unmodifiable list.
+     *
+     * @param <E>  type of elements
+     */
+    @SafeVarargs
+    @SuppressWarnings({"unchecked", "varargs"})
+    private static <E> List<E> toList(E... elements) {
+        return Collections.unmodifiableList(Arrays.asList(elements));
```

Comment:
Is a defensive copy needed here?
The caller of `newStringTemplate` could retain a reference to the Object[] array and modify it later.

Reply:
Alternatively, `SharedSecrets.getJavaUtilCollectionAccess().listFromTrustedArrayNullsAllowed(elements)` (or `elements.clone()`) could be used here, to reduce amount of wrappers.

Reply:
The `newStringTemplate` methods are internal and elements is safe. Will get rid of the generics and varargs.  

@amaembo I was using listFromTrustedArrayNullsAllowed early on but ran into an edge case (crash). Will try again.